### PR TITLE
TraceId in LAPI [DPP-1372].

### PR DESCRIPTION
**Pull request by atriantafyllos-da**
[DPP-1372](https://digitalasset.atlassian.net/browse/DPP-1372)

[Design doc](https://docs.google.com/document/d/10X-fbC2MuLI2wD5xe-jvwrU25G8t5DGK0_yve-NqPvw/edit?pli=1#heading=h.j1o9vy5fqmrz)

Currently ledger-api server lacks the trace id which is used extensively in canton to trace the requests and appears in the logfiles of canton consistently. The trace-id is derived by the open telemetry context.

As described in [design-doc](https://docs.google.com/document/d/10X-fbC2MuLI2wD5xe-jvwrU25G8t5DGK0_yve-NqPvw/edit?pli=1#heading=h.j1o9vy5fqmrz) currently the ledger-api server logs are matched with the canton logs only by a rosetta-stone log entry which states both the trace-id from canton and submission-id from ledger-api server. With this PR, the trace-id is passed via the grpc context to ledger-api server and identifying of log entries from both codebases is happening with the use of the trace-id.

[DPP-1372]: https://digitalasset.atlassian.net/browse/DPP-1372?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala [hunk 1]
```diff
@@ -79,15 +79,15 @@ private[apiserver] final class ApiSubmissionService private[services] (
...
+)(implicit executionContext: ExecutionContext)
     extends CommandSubmissionService
     with AutoCloseable {
 
   private val logger = ContextualizedLogger.get(this.getClass)
 
   override def submit(
       request: SubmitRequest
-  )(implicit telemetryContext: TelemetryContext): Future[Unit] =
+  )(implicit telemetryContext: TelemetryContext, loggingContext: LoggingContext): Future[Unit] = {
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala:90]: you can leave it without `{` right?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala [hunk 2]
```diff
@@ -44,42 +46,47 @@ final class ApiParticipantPruningService private (
...
-              _ <- pruneWriteService(pruneUpTo, submissionId, request.pruneAllDivulgedContracts)
-
-              pruneResponse <- pruneLedgerApiServerIndex(
-                pruneUpTo,
-                request.pruneAllDivulgedContracts,
-              )
+    val telemetryContext: TelemetryContext =
+      DefaultTelemetry.contextFromGrpcThreadLocalContext()
+    withEnrichedLoggingContext(logging.traceId(telemetryContext.traceId)) {
+      implicit loggingContext =>
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala:52]: This 4-line pattern is repeated at least 10 times throughout the file. How about we extract it in a utils method and use that one? Aside from the repetition, I think less verbosity in these API methods would be nice as well

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/logging/package.scala [hunk 3]
```diff
@@ -91,6 +91,12 @@ package object logging {
   private[services] def workflowId(id: String): LoggingEntry =
     "workflowId" -> id
 
+  private[services] def packageId(id: String): LoggingEntry =
+    "packageId" -> id
+
+  private[services] def traceId(id: Option[String]): LoggingEntry =
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/logging/package.scala:97]: What's the difference between this one and the other one that you just introduced [here](https://github.com/digital-asset/daml/pull/15995/files#diff-c877e2c32219d2bbf486172296c6af3b44490410f1bdd52db0163a76f911c8f4R10) ?

File: observability/telemetry/src/main/scala/com/daml/telemetry/TelemetryContext.scala [hunk 4]
```diff
@@ -168,6 +170,11 @@ protected class DefaultTelemetryContext(protected val tracer: Tracer, protected
   }
 
   override def openTelemetryContext: Context = Context.current.`with`(span)
+
+  lazy val traceId: Option[String] = Option(Span.fromContextOrNull(this.openTelemetryContext))
+    .filter(_.getSpanContext.isValid)
+    .map(_.getSpanContext.getTraceId)
```
Comment (tudor-da) [inline observability/telemetry/src/main/scala/com/daml/telemetry/TelemetryContext.scala:176]: Is this the traceId that is used in Canton as well? It could help if you could add a brief description of what you aim to solve here to the PR or to the ticket otherwise it's harder to review.

**Comment by tudor-da** [comment]
The PR generally looks good to me, but I'll wait with the approve until you add some description which can guide the review.

File: observability/telemetry/src/main/scala/com/daml/telemetry/TelemetryContext.scala (hunk 4 above)
Comment (atriantafyllos-da) [reply observability/telemetry/src/main/scala/com/daml/telemetry/TelemetryContext.scala:176]: Updated the description.

Comment (tudor-da) [reply observability/telemetry/src/main/scala/com/daml/telemetry/TelemetryContext.scala:176]: Thanks! I see we already have a `Span` in scope (argument to the `DefaultTelemetryContext`). Why don't we use this one? Also, is there a specific reason on why we want to have `traceId` lazy?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/logging/package.scala (hunk 3 above)
Comment (atriantafyllos-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/logging/package.scala:97]: You are right I am removing the duplication for `traceId`.

Comment (atriantafyllos-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/logging/package.scala:97]: Or should I move all of them there?

File: observability/telemetry/src/main/scala/com/daml/telemetry/TelemetryContext.scala (hunk 4 above)
Comment (atriantafyllos-da) [reply observability/telemetry/src/main/scala/com/daml/telemetry/TelemetryContext.scala]: You are right I will use the existent `Span`.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala (hunk 2 above)
Comment (atriantafyllos-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala]: replaced w/ :

https://github.com/digital-asset/daml/pull/15995/files#diff-734b593ac0044ef8aa4560732b8b405e54921f1c526421a97515cffd35e9ae51R52-R53

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/logging/package.scala (hunk 3 above)
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/logging/package.scala]: I would do what Tudor has suggested, i.e. one definition of traceId in common, other definitions remain where they are

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_15/InterfaceSubscriptionsIT.scala [hunk 5]
```diff
@@ -481,8 +481,15 @@ class InterfaceSubscriptionsIT extends LedgerTestSuite {
       )
       assertEquals(
         "1 and 3 produce the same views (but not the same create arguments)",
-        transactions1.map(updateTransaction()),
-        transactions3.map(updateTransaction(emptyContractKey = true, emptyCreateArguments = true)),
+        // do not check on details since tid is contained and it is expected to be different
+        transactions1.map(updateTransaction(emptyDetails = true)),
```
Comment (mziolekda) [inline ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_15/InterfaceSubscriptionsIT.scala:485]: I don't understand why the traceId related changes blew up in this place. It seems unrelated to the subject.

Comment (atriantafyllos-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_15/InterfaceSubscriptionsIT.scala:485]: The details of the status of the interface views contains the tid:
```
DAML_INTERPRETATION_ERROR
participant'participant1')
tid"'9463676f15dff8cebe9b2f3599be8baa'
category9
definite_answerfalse
```
and so for different transactions the tids differ.

Comment (tudor-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_15/InterfaceSubscriptionsIT.scala:485]: On failure, the interface view grpc `Status` is constructed by means of the self-service error code (which inlines property map of the `LoggingContext`, which has the `traceId` now) - see https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/main/scala/platform/store/dao/events/LfValueTranslation.scala#L437-L450

Comment (tudor-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_15/InterfaceSubscriptionsIT.scala:485]: I don't see why we should remove it, but I think your approach with ignoring the status details is fine 👍

Comment (mziolekda) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_15/InterfaceSubscriptionsIT.scala:485]: It is absolutely fine. It just exposes the fact that our current method of asserting on the error message is extremely brittle

**Review by mziolekda — APPROVED** [review]

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/logging/package.scala [hunk 6]
```diff
@@ -91,6 +91,9 @@ package object logging {
   private[services] def workflowId(id: String): LoggingEntry =
     "workflowId" -> id
 
+  private[services] def packageId(id: String): LoggingEntry =
+    "packageId" -> id
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/logging/package.scala:95]: great : 👍

**Review by tudor-da — APPROVED** [review]
